### ISSUE TITLE
Potential race condition causing resolver to throw error indefinitely

### ISSUE THREAD

**Issue by louischan-oursky** [authgear/authgear-server#2954]
```
time="2023-02-16T13:33:17Z" level=error msg="panic occurred" error="runtime error: invalid memory address or nil pointer dereference" logger=panic-middleware stack="[log.PanicValue at /src/pkg/util/log/panic.go:56 middleware.(*PanicMiddleware).Handle.func1.3 at /src/pkg/lib/infra/middleware/panic.go:48 runtime.gopanic at /usr/local/go/src/runtime/panic.go:884 runtime.panicmem at /usr/local/go/src/runtime/panic.go:260 runtime.sigpanic at /usr/local/go/src/runtime/signal_unix.go:835 deps.(*RootProvider).NewAppProvider at /src/pkg/lib/deps/providers.go:92 deps.(*RequestMiddleware).Handle.func1 at /src/pkg/lib/deps/middleware_request.go:37 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 httputil.StaticSecurityHeaders.func1 at /src/pkg/util/httputil/static_security_headers.go:10 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 middleware.(*SentryMiddleware).Handle.func1 at /src/pkg/lib/infra/middleware/sentry.go:22 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 middleware.(*BodyLimitMiddleware).Handle.func1 at /src/pkg/lib/infra/middleware/body_limit.go:16 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 middleware.(*PanicMiddleware).Handle.func1 at /src/pkg/lib/infra/middleware/panic.go:63 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 httprouter.(*Router).Handler.func1 at /go/pkg/mod/github.com/julienschmidt/httprouter@v1.3.0/router.go:275 httprouter.(*Router).ServeHTTP at /go/pkg/mod/github.com/julienschmidt/httprouter@v1.3.0/router.go:387 httproute.(*Router).ServeHTTP at /src/pkg/util/httproute/httproute.go:85 http.serverHandler.ServeHTTP at /usr/local/go/src/net/http/server.go:2947 http.(*conn).serve at /usr/local/go/src/net/http/server.go:1991]"
```

**Comment by louischan-oursky** [comment]
Happened again on 2023-04-18 on staging

```
time="2023-04-18T07:31:21Z" level=error msg="panic occurred" error="hook-handle: transaction not started" logger=panic-middleware stack="[log.PanicValue at /src/pkg/util/log/panic.go:56 middleware.(*PanicMiddleware).Handle.func1.3 at /src/pkg/lib/infra/middleware/panic.go:48 runtime.gopanic at /usr/local/go/src/runtime/panic.go:884 db.(*HookHandle).WithTx.func1 at /src/pkg/lib/infra/db/hook_handle.go:88 runtime.gopanic at /usr/local/go/src/runtime/panic.go:884 db.(*HookHandle).conn at /src/pkg/lib/infra/db/hook_handle.go:36 db.(*SQLExecutor).QueryRowWith at /src/pkg/lib/infra/db/sql_executor.go:59 configsource.(*Store).GetDatabaseSourceByAppID at /src/pkg/lib/config/configsource/store.go:86 configsource.(*dbApp).doLoad.func1 at /src/pkg/lib/config/configsource/database.go:324 db.(*HookHandle).WithTx at /src/pkg/lib/infra/db/hook_handle.go:110 configsource.(*dbApp).doLoad at /src/pkg/lib/config/configsource/database.go:322 configsource.(*dbApp).Load.func1 at /src/pkg/lib/config/configsource/database.go:313 sync.(*Once).doSlow at /usr/local/go/src/sync/once.go:74 sync.(*Once).Do at /usr/local/go/src/sync/once.go:65 configsource.(*dbApp).Load at /src/pkg/lib/config/configsource/database.go:312 configsource.(*Database).ResolveContext at /src/pkg/lib/config/configsource/database.go:176 configsource.(*ConfigSource).ProvideContext at /src/pkg/lib/config/configsource/source.go:33 deps.(*RequestMiddleware).Handle.func1 at /src/pkg/lib/deps/middleware_request.go:26 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 httputil.StaticSecurityHeaders.func1 at /src/pkg/util/httputil/static_security_headers.go:10 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 middleware.(*SentryMiddleware).Handle.func1 at /src/pkg/lib/infra/middleware/sentry.go:22 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 middleware.(*BodyLimitMiddleware).Handle.func1 at /src/pkg/lib/infra/middleware/body_limit.go:16 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 middleware.(*PanicMiddleware).Handle.func1 at /src/pkg/lib/infra/middleware/panic.go:63 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 httprouter.(*Router).Handler.func1 at /go/pkg/mod/github.com/julienschmidt/httprouter@v1.3.0/router.go:275 httprouter.(*Router).ServeHTTP at /go/pkg/mod/github.com/julienschmidt/httprouter@v1.3.0/router.go:387 httproute.(*Router).ServeHTTP at /src/pkg/util/httproute/httproute.go:85 http.serverHandler.ServeHTTP at /usr/local/go/src/net/http/server.go:2947 http.(*conn).serve at /usr/local/go/src/net/http/server.go:1991]"
time="2023-04-18T07:31:21Z" level=error msg="panic occurred" error="runtime error: invalid memory address or nil pointer dereference" logger=panic-middleware stack="[log.PanicValue at /src/pkg/util/log/panic.go:56 middleware.(*PanicMiddleware).Handle.func1.3 at /src/pkg/lib/infra/middleware/panic.go:48 runtime.gopanic at /usr/local/go/src/runtime/panic.go:884 runtime.panicmem at /usr/local/go/src/runtime/panic.go:260 runtime.sigpanic at /usr/local/go/src/runtime/signal_unix.go:835 deps.(*RootProvider).NewAppProvider at /src/pkg/lib/deps/providers.go:92 deps.(*RequestMiddleware).Handle.func1 at /src/pkg/lib/deps/middleware_request.go:37 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 httputil.StaticSecurityHeaders.func1 at /src/pkg/util/httputil/static_security_headers.go:10 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 middleware.(*SentryMiddleware).Handle.func1 at /src/pkg/lib/infra/middleware/sentry.go:22 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 middleware.(*BodyLimitMiddleware).Handle.func1 at /src/pkg/lib/infra/middleware/body_limit.go:16 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 middleware.(*PanicMiddleware).Handle.func1 at /src/pkg/lib/infra/middleware/panic.go:63 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 httprouter.(*Router).Handler.func1 at /go/pkg/mod/github.com/julienschmidt/httprouter@v1.3.0/router.go:275 httprouter.(*Router).ServeHTTP at /go/pkg/mod/github.com/julienschmidt/httprouter@v1.3.0/router.go:387 httproute.(*Router).ServeHTTP at /src/pkg/util/httproute/httproute.go:85 http.serverHandler.ServeHTTP at /usr/local/go/src/net/http/server.go:2947 http.(*conn).serve at /usr/local/go/src/net/http/server.go:1991]"
time="2023-04-18T07:31:21Z" level=error msg="panic occurred" error="runtime error: invalid memory address or nil pointer dereference" logger=panic-middleware stack="[log.PanicValue at /src/pkg/util/log/panic.go:56 middleware.(*PanicMiddleware).Handle.func1.3 at /src/pkg/lib/infra/middleware/panic.go:48 runtime.gopanic at /usr/local/go/src/runtime/panic.go:884 runtime.panicmem at /usr/local/go/src/runtime/panic.go:260 runtime.sigpanic at /usr/local/go/src/runtime/signal_unix.go:835 deps.(*RootProvider).NewAppProvider at /src/pkg/lib/deps/providers.go:92 deps.(*RequestMiddleware).Handle.func1 at /src/pkg/lib/deps/middleware_request.go:37 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 httputil.StaticSecurityHeaders.func1 at /src/pkg/util/httputil/static_security_headers.go:10 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 middleware.(*SentryMiddleware).Handle.func1 at /src/pkg/lib/infra/middleware/sentry.go:22 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 middleware.(*BodyLimitMiddleware).Handle.func1 at /src/pkg/lib/infra/middleware/body_limit.go:16 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 middleware.(*PanicMiddleware).Handle.func1 at /src/pkg/lib/infra/middleware/panic.go:63 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 httprouter.(*Router).Handler.func1 at /go/pkg/mod/github.com/julienschmidt/httprouter@v1.3.0/router.go:275 httprouter.(*Router).ServeHTTP at /go/pkg/mod/github.com/julienschmidt/httprouter@v1.3.0/router.go:387 httproute.(*Router).ServeHTTP at /src/pkg/util/httproute/httproute.go:85 http.serverHandler.ServeHTTP at /usr/local/go/src/net/http/server.go:2947 http.(*conn).serve at /usr/local/go/src/net/http/server.go:1991]"
time="2023-04-18T07:31:21Z" level=error msg="panic occurred" error="runtime error: invalid memory address or nil pointer dereference" logger=panic-middleware stack="[log.PanicValue at /src/pkg/util/log/panic.go:56 middleware.(*PanicMiddleware).Handle.func1.3 at /src/pkg/lib/infra/middleware/panic.go:48 runtime.gopanic at /usr/local/go/src/runtime/panic.go:884 runtime.panicmem at /usr/local/go/src/runtime/panic.go:260 runtime.sigpanic at /usr/local/go/src/runtime/signal_unix.go:835 deps.(*RootProvider).NewAppProvider at /src/pkg/lib/deps/providers.go:92 deps.(*RequestMiddleware).Handle.func1 at /src/pkg/lib/deps/middleware_request.go:37 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 httputil.StaticSecurityHeaders.func1 at /src/pkg/util/httputil/static_security_headers.go:10 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 middleware.(*SentryMiddleware).Handle.func1 at /src/pkg/lib/infra/middleware/sentry.go:22 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 middleware.(*BodyLimitMiddleware).Handle.func1 at /src/pkg/lib/infra/middleware/body_limit.go:16 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 middleware.(*PanicMiddleware).Handle.func1 at /src/pkg/lib/infra/middleware/panic.go:63 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 httprouter.(*Router).Handler.func1 at /go/pkg/mod/github.com/julienschmidt/httprouter@v1.3.0/router.go:275 httprouter.(*Router).ServeHTTP at /go/pkg/mod/github.com/julienschmidt/httprouter@v1.3.0/router.go:387 httproute.(*Router).ServeHTTP at /src/pkg/util/httproute/httproute.go:85 http.serverHandler.ServeHTTP at /usr/local/go/src/net/http/server.go:2947 http.(*conn).serve at /usr/local/go/src/net/http/server.go:1991]"
time="2023-04-18T07:31:21Z" level=error msg="panic occurred" error="runtime error: invalid memory address or nil pointer dereference" logger=panic-middleware stack="[log.PanicValue at /src/pkg/util/log/panic.go:56 middleware.(*PanicMiddleware).Handle.func1.3 at /src/pkg/lib/infra/middleware/panic.go:48 runtime.gopanic at /usr/local/go/src/runtime/panic.go:884 runtime.panicmem at /usr/local/go/src/runtime/panic.go:260 runtime.sigpanic at /usr/local/go/src/runtime/signal_unix.go:835 deps.(*RootProvider).NewAppProvider at /src/pkg/lib/deps/providers.go:92 deps.(*RequestMiddleware).Handle.func1 at /src/pkg/lib/deps/middleware_request.go:37 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 httputil.StaticSecurityHeaders.func1 at /src/pkg/util/httputil/static_security_headers.go:10 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 middleware.(*SentryMiddleware).Handle.func1 at /src/pkg/lib/infra/middleware/sentry.go:22 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 middleware.(*BodyLimitMiddleware).Handle.func1 at /src/pkg/lib/infra/middleware/body_limit.go:16 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 middleware.(*PanicMiddleware).Handle.func1 at /src/pkg/lib/infra/middleware/panic.go:63 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 httprouter.(*Router).Handler.func1 at /go/pkg/mod/github.com/julienschmidt/httprouter@v1.3.0/router.go:275 httprouter.(*Router).ServeHTTP at /go/pkg/mod/github.com/julienschmidt/httprouter@v1.3.0/router.go:387 httproute.(*Router).ServeHTTP at /src/pkg/util/httproute/httproute.go:85 http.serverHandler.ServeHTTP at /usr/local/go/src/net/http/server.go:2947 http.(*conn).serve at /usr/local/go/src/net/http/server.go:1991]"
time="2023-04-18T07:31:21Z" level=error msg="panic occurred" error="runtime error: invalid memory address or nil pointer dereference" logger=panic-middleware stack="[log.PanicValue at /src/pkg/util/log/panic.go:56 middleware.(*PanicMiddleware).Handle.func1.3 at /src/pkg/lib/infra/middleware/panic.go:48 runtime.gopanic at /usr/local/go/src/runtime/panic.go:884 runtime.panicmem at /usr/local/go/src/runtime/panic.go:260 runtime.sigpanic at /usr/local/go/src/runtime/signal_unix.go:835 deps.(*RootProvider).NewAppProvider at /src/pkg/lib/deps/providers.go:92 deps.(*RequestMiddleware).Handle.func1 at /src/pkg/lib/deps/middleware_request.go:37 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 httputil.StaticSecurityHeaders.func1 at /src/pkg/util/httputil/static_security_headers.go:10 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 middleware.(*SentryMiddleware).Handle.func1 at /src/pkg/lib/infra/middleware/sentry.go:22 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 middleware.(*BodyLimitMiddleware).Handle.func1 at /src/pkg/lib/infra/middleware/body_limit.go:16 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 middleware.(*PanicMiddleware).Handle.func1 at /src/pkg/lib/infra/middleware/panic.go:63 http.HandlerFunc.ServeHTTP at /usr/local/go/src/net/http/server.go:2109 httprouter.(*Router).Handler.func1 at /go/pkg/mod/github.com/julienschmidt/httprouter@v1.3.0/router.go:275 httprouter.(*Router).ServeHTTP at /go/pkg/mod/github.com/julienschmidt/httprouter@v1.3.0/router.go:387 httproute.(*Router).ServeHTTP at /src/pkg/util/httproute/httproute.go:85 http.serverHandler.ServeHTTP at /usr/local/go/src/net/http/server.go:2947 http.(*conn).serve at /usr/local/go/src/net/http/server.go:1991]"
```

**Comment by kiootic** [comment]
`configsource.Database` contains a map of `configsource.dbApp`. Considering `dbApp.doLoad`, it is thread-safe with internal state of `dbApp`, but calling to `Database.WithTx` is not safe since the `*db.HookHandle` is shared across all instances of `dbApp`.

We may need to guard `doLoad` with a lock, or to make `WithTx` run with a cloned handle.

**Comment by louischan-oursky** [comment]
Or we can change globaldb.Handle not to depend on db.HookHandle. Then withTx should be thread-safe.

**Comment by louischan-oursky** [comment]
Turns out it is not trivial. SQLExecutor expects Handle.conn() to be present.

**Comment by louischan-oursky** [comment]
Per discussion, we would like to try the following approach

configsource.Database has two responsibilities. It provides caching and loading the config of an app. We would like to see if we can separate the two responsibilities. The loading should be in request-scope, so that the HookHandle is bound to the request, eliminating the race condition.

**Comment by tung2744** [comment]
@louischan-oursky Do you have any method to trigger it in local?

**Comment by louischan-oursky** [comment]
No :(

**Comment by fungc-io** [comment]
@louischan-oursky is this fixed in `2023C01` branch?

**Comment by louischan-oursky** [comment]
Yes

**Comment by louischan-oursky** [comment]
Yes

**Comment by louischan-oursky** [comment]
Yes

**Comment by fungc-io** [comment]
Fixed in 2023C01 branch, pending merge into main
